test/org/centos/contra/Infra/Utils.groovy

```diff
@@ -0,0 +1,369 @@
+    }
+
+    def "creating beaker instances drops to defaults when data is missing"() {
+        setup:
+            def beaker = [instances: [[:]]]
+        when:
+            def instances = infraUtils.createBeakerInstances(beaker as HashMap)
+
+        then:
+            instances != null
+    }
+
+    def "create openstack instances"() {
+        setup:
+            def openstack = config.infra.provision.cloud.openstack
+            def len = openstack.instances.size()
+            def i = 0
+        when:
+            def instances = infraUtils.createOpenstackInstances(openstack as HashMap)
+
+        then:
+            instances != null
+            instances.every { instance -> instance.getNetwork() == openstack.network[0] }
+            instances.every { instance -> instance.getKeyPair() == openstack.key_pair }
+            instances.every { instance -> instance.getSecurityGroups() == openstack.security_groups.join(', ') }
+            instances.every { instance -> instance.getName() == openstack.instances[i++].name }
+            instances.every { instance -> instance.getFlavor() == openstack.instances[len - (i--)].flavor }
+            instances.every { instance -> instance.getImage() == openstack.instances[i++].image }
+            instances.every { instance -> instance.getFipPool() == openstack.instances[len - (i--)].floating_ip_pool }
+            instances.every { instance -> instance.getUser() == openstack.instances[i++].user }
+    }
+
+    def "creating openstack instances drops to defaults when data is missing"() {
+        setup:
+            def openstack = [instances: [[:]]]
+        when:
+            def instances = infraUtils.createOpenstackInstances(openstack as HashMap)
+
+        then:
+            instances != null
+    }
+
+    def "able to execute linchpin"() {
+        setup:
+            getPipelineMock("env.getEnvironment")() >> []
+
+        when:
+            infraUtils.executeInLinchpin("validate", "", true, "abc")
+
+        then:
+            noExceptionThrown()
+    }
+
+    def "ansible containers can be called without errors"() {
+        setup:
+            getPipelineMock("env.getEnvironment")() >> []
+
+        when:
+            infraUtils.executeInAnsible("anything.yml", "", true, "abc")
+
+        then:
+            noExceptionThrown()
+    }
+
+    def "executeInAnsible sends correct command to the container for playbook execution"() {
+        setup:
+            infraUtils.getBinding().setVariable("WORKSPACE", "workspace")
+        when:
+            def status = infraUtils.executeInAnsible("anything.yml", "", true, "abc")
+        then:
+            1 * getPipelineMock("sh")("""ansible-playbook -vvv -i "workspace/inventory" "workspace/anything.yml" """)
+    }
+
+    def "aws topology is created and is correct"() {
+        setup:
+            getPipelineMock("libraryResource")(_) >> {
+                return '''${index}_${providerType}:
```

Comment:
Im wondering if we could put this mock topologies into the resources directory and load them in. It would shrink this file and clean it up a little. Thoughts?

Reply:
We could certainly clean up the test by putting them in the test directory. I'd really love to be able to pull them from the resources directory, though! They already exist in the resources directory, and in fact, by duplicating them here (or in the test directory if we do that), we are missing some key test functionality, since someone could break the actual templates and the test wouldn't notice. But nothing I've tried has successfully allowed me to load anything from the resources directory within a test. I've spent a lot of hours attempting that. If you have any ideas about what might allow us to pull them from the resources directory, I'm all ears.